src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs

```diff
@@ -911,5 +913,114 @@ private void ValidateTPHMapping(IEntityType rootEntityType, bool forTables)
 
         private static string Format(string tableName, string schema)
             => schema == null ? tableName : schema + "." + tableName;
+
+        /// <summary>
+        ///     Validates that the properties of any one index are
+        ///     all mapped to columns on at least one common table.
+        /// </summary>
+        /// <param name="model"> The model to validate. </param>
+        /// <param name="logger"> The logger to use. </param>
+        protected virtual void ValidateIndexProperties(
+            [NotNull] IModel model, [NotNull] IDiagnosticsLogger<DbLoggerCategory.Model.Validation> logger)
+        {
+            Check.NotNull(model, nameof(model));
+
+            foreach (var entityType in model.GetEntityTypes())
+            {
+                foreach (var index in entityType.GetIndexes()
+                    .Where(i => ConfigurationSource.Convention != ((IConventionIndex)i).GetConfigurationSource()))
+                {
+                    IProperty propertyNotMappedToAnyTable = null;
+                    Tuple<string, List<(string Table, string Schema)>> firstPropertyTables = null;
+                    Tuple<string, List<(string Table, string Schema)>> lastPropertyTables = null;
+                    HashSet<(string Table, string Schema)> overlappingTables = null;
+                    foreach (var property in index.Properties)
+                    {
+                        var tablesMappedToProperty = property.DeclaringEntityType.GetDerivedTypesInclusive()
```

Comment:
```suggestion
                        var tablesMappedToProperty = entityType.GetDerivedTypesInclusive()
```

We should only look at the entity types where the index is present.

Reply:
Either this or the comment below about `GetDeclaredIndexes()` causes the test Detects_index_properties_mapped_to_different_tables_in_TPT_hierarchy to fail. I'll have to look into this.

Reply:
I see. You can leave it as is then.

Reply:
We'll need to do something different here for TPC, but we'll see when we get there.